### PR TITLE
Lockless LintStore

### PR DESCRIPTION
**Pull request by Mark-Simulacrum**
This removes mutability from the lint store after registration. Each commit stands alone, for the most part, though they don't make sense out of sequence.

The intent here is to move LintStore to a more parallel-friendly architecture, although also just a cleaner one from an implementation perspective. Specifically, this has the following changes:
 * We no longer implicitly register lints when registering lint passes
    * For the most part this means that registration calls now likely want to call something like:
       `lint_store.register_lints(&Pass::get_lints())` as well as `register_*_pass`.
    * In theory this is a simplification as it's much easier for folks to just register lints and then have passes that implement whichever lint however they want, rather than necessarily tying passes to lints.
 * Lint passes still have a list of associated lints, but a followup PR could plausibly change that
   * This list must be known for a given pass type, not instance, i.e., `fn get_lints()` is the signature instead of `fn get_lints(&self)` as before.
 * We do not store pass objects, instead storing constructor functions. This means we always get new passes when running lints (this happens approximately once though for a given compiler session, so no behavior change is expected).
 * Registration API is _much_ simpler: generally all functions are just taking `Fn() -> PassObject` rather than several different `bool`s.

### CODE REVIEW

**Comment by Manishearth** [comment]
We should hold off on landing this until clippy's recentmost rustup lands, fwiw

(i.e. only land this if clippy is toolstate green, because if this breaks things in an unfixable way we'd want to be able to deal with things immediately)

**Comment by Manishearth** [comment]
Won't have time to do a proper review but what you describe in the PR seems fine from a clippy POV.

Tying passes to lints, weakly, may still be useful because it would be nice to selectively run passes based on what lints are allowed (provided those passes are pure, of course)

**Comment by Mark-Simulacrum** [comment]
I'm happy to try and help get a PR going on the clippy side to fix this, FWIW, though I've never contributed a patch to clippy. It also looks like it'll break RLS based on some quick grepping, but not quite sure why there.

Would that be helpful?

**Comment by Mark-Simulacrum** [comment]
> Tying passes to lints, weakly, may still be useful because it would be nice to selectively run passes based on what lints are allowed (provided those passes are pure, of course)

I think that might be an interesting optimization in the future -- these days it's not something we do -- but this PR does not prevent such work from happening if/when it becomes desirable.

**Comment by Manishearth** [comment]
If you want to make a PR for clippy that would be great, of course! I'm not too worried about the breakage from this: it sounds like a lot of mechanical fixes, but I won't say no to help :smile: 

(As clippy toolstate is broken right now you'll have to graft in the rustup PR first)

**Comment by Manishearth** [comment]
> I think that might be an interesting optimization in the future -- these days it's not something we do -- but this PR does not prevent such work from happening if/when it becomes desirable.

Yep, this was more of a note than a blocker

File: src/librustc/lint/context.rs
Comment (Zoxc) [inline src/librustc/lint/context.rs]: Module passes will be created once per module, so this comment isn't true for them.

**Comment by Zoxc** [comment]
I'm not sure about `Duplicate lint specifications are always bug!`, but the rest of the commits looks good to me.

File: src/librustc/lint/context.rs
Comment (Mark-Simulacrum) [reply src/librustc/lint/context.rs]: Ah, yeah, I'll leave a comment to that effect. We don't actually do that today anyway I believe -- at least, the previous code never had an impl of `fresh_late_pass`, it always panic'd...

Comment (Zoxc) [reply src/librustc/lint/context.rs]: All the late module passes are hardcoded in the compiler so they don't appear in this list, hence they don't need a real impl of `fresh_late_pass`, but I guess we shouldn't allow plugins to add to `late_module_passes` either since that might not won't work for incremental compilation.

**Comment by mati865** [comment]
> We should hold off on landing this until clippy's recentmost rustup lands, fwiw

Already landed on Clippy side and should soon land in Rust: https://github.com/rust-lang/rust/pull/65206

**Comment by Mark-Simulacrum** [comment]
Pushed up new commits moving future compatibility information to lints directly as another step in moving us off of LintStore on Session -- will try to spend some time tomorrow finishing the decoupling entirely.

**Comment by Mark-Simulacrum** [comment]
And, with the last set of commits, this PR now entirely drops LintStore as a mutable (via locks) field from global-ish structures (Session, in particular).

Plugin registration of lints is also altered; they are now expected to directly call methods of LintStore rather than pushing into vectors and then those vectors being appended onto lint store. This seems both easier and less error prone.

File: src/librustc/session/mod.rs
Comment (Zoxc) [inline src/librustc/session/mod.rs]: Keep this fixme for `buffered_lints`?

File: src/librustc/ty/context.rs
Comment (Zoxc) [inline src/librustc/ty/context.rs]: Should be `Lrc`.

File: src/librustc_driver/lib.rs
Comment (Zoxc) [inline src/librustc_driver/lib.rs]: This seems like it should be refactored into a `LintStore` constructor.

File: src/librustc_plugin/registry.rs
Comment (Zoxc) [inline src/librustc_plugin/registry.rs]: Why did you remove these functions? This seems to break plugins registering lints?

Comment (Mark-Simulacrum) [reply src/librustc_plugin/registry.rs]: We publicly expose the lint store now, and it should be a simple transform to use `registry.lint_store.register` vs these functions. Plugins needed to change anyway as we've changed the signature here and expect separate registration of lints and lint passes.

**Comment by Mark-Simulacrum** [comment]
Okay I've added one more commit that makes clippy changes easier and filed a PR on clippy -- https://github.com/rust-lang/rust-clippy/pull/4650 -- that makes the relevant changes there as well.

**Comment by Mark-Simulacrum** [comment]
@Manishearth Not sure if you will have time to review this, if not, please reassign (maybe to @Zoxc as they've also done some review here already).

I believe the pieces here are all ready to go pending this try build -- https://github.com/rust-lang/rust-clippy/pull/4650 is up with the majority of the clippy changes.

**Comment by Manishearth** [comment]
Oh, yeah, I mentioned earlier that I won't have time to do a proper review.

The general shape of things looks good to me, and i'll probably go through this eventually, but yeah, worth finding a different reviewer.

**Comment by Mark-Simulacrum** [comment]
In that case, let's try r? @Zoxc and maybe @estebank would be able to as well

File: src/librustc/lint/builtin.rs
Comment (nikomatsakis) [inline src/librustc/lint/builtin.rs]: we should probably update [the page on this in the forge](https://forge.rust-lang.org/compiler/bug-fix-procedure.html#issuing-future-compatibility-warnings) -- actually, that content should really move to rustc-guide I expect -- or at least some of it

File: src/librustc/lint/context.rs
Comment (nikomatsakis) [inline src/librustc/lint/context.rs]: Nit: put `&mut self` on next line, to align with rustfmt? (Similarly, the `) {` on the next line)

File: src/librustc_driver/lib.rs
Comment (nikomatsakis) [inline src/librustc_driver/lib.rs]: Nit: doc-comment

File: src/librustc_plugin/registry.rs
Comment (nikomatsakis) [inline src/librustc_plugin/registry.rs]: Nit: doc-comment, I think something simple like this would suffice

---

The `LintStore` permits plugins to register new lints.

File: src/librustc_interface/interface.rs
Comment (nikomatsakis) [inline src/librustc_interface/interface.rs]: It'd be nice to have a comment here. Something like:

---

Callback that is invoked during lint registration. Use the `&mut LintStore` value to register new lints with the compiler as needed.

**Review by nikomatsakis — APPROVED** [review]
This new design is really quite nice. Simpler overall, and lock-free to boot. Huzzah!

**Comment by nikomatsakis** [comment]
r=me with nits addressed

**Pull request merged**
